"""
Snakefile to clean and join illumina reads for read-based analysis

We are working from the old recipe from illuminaPrep:
 * trimmomatic to remove primers
 * pandaseq to join and screen
 * trimmomatic again to trim ends and windows

This is in the testing phase:
 * we have multiple candidate tools for each step
 * this workflow will link them together based on output file names

"""
import re
from python.common import get_version, add_stats_outputs
from python.qc import setup_qc_outputs
from snakemake.logging import logger

# parses config to setup outputs
outputs = config.setdefault('outputs',set())
outputs.update(setup_qc_outputs(config))

include: "common/transitions.snake"

# all three joining programs are interchangeable
joining_program = config.get('joining_program','pear')
include: 'qc/{joining_program}.snake'.format(**vars())

logger.debug(config)

# add my tools to the path (Somewhat of a hack, maybe I'll get my stuff in
# conda eventually)
snakefile_path = os.path.dirname(os.path.abspath(workflow.snakefile))
pymg_dir = os.path.sep.join([snakefile_path, 'tools', 'pymg'])
batch_dir = os.path.sep.join([snakefile_path, 'tools', 'batch'])
os.environ['PATH'] += os.pathsep + os.pathsep.join([pymg_dir, batch_dir])
logger.debug(os.environ['PATH'])
#

# adds fastq stats files to output (only if config[run_stats] is set to True
if config.get('run_stats', False) in [True, 'True']:
    add_stats_outputs(workflow.snakefile, config) 
    include: 'common/stats.snake'

# Tell snakemake what to make
rule clean_illumina_outputs:
    input:
        outputs

