"""
Annotate reads using six frame translation. Lastal has this built in (aka: lastx) when given the -F flag. For hmmsearch, we'll run through transeq 1st
"""

import sys
import os
snakefile_path=os.path.dirname(os.path.abspath(workflow.snakefile))
sys.path[0]=os.path.join(snakefile_path,'..','python')
from common import get_version
from annotate import get_db_assignment_params,get_db_types,\
                     get_hit_table_name_from_wildcards_db
#configfile: "annotation.yaml"
#########
# CONFIGURATION
#
# rank to collect taxa on (defaults to order)
config.setdefault('clade_ranks', ['order'])
#
# threads per search are set in the threads: parameter
# Either use the db name or "default". EG:
# threads:
#   default: 5
#   lastal: 20
# Defaults to 8 for eveerything
default_threads=config.setdefault('threads',{}).get('default', 3)
#
# pull out taxdb (usually refseq) and list of gene family dbs
gene_family_dbs, config['taxdb'] = get_db_types(config)
#
config.setdefault('annotation_hit_table_map',{})
# End configuration
##########

# command for removing transeq readingframe suffixes from hit names
translation_filter = "perl -pe 's/^(\S+)_\d+(\s+)/\\1\\2/'"

include: "common.snake"

rule remove_colons:
    " transeq doesn't like colons "
    input: "{prefix}.{ext}"
    output: "{prefix}.nocolon.{ext,f[^.]+}"
    benchmark: "benchmarks/{prefix}.nocolon.{ext}.time"
    shell:
        "cat {input} \
         | perl -pe 'if (m/^>/) {{ tr/:/_/; }}' \
         > {output}"

rule transeq:
    " translate in six frames "
    input: "{prefix}.fasta"
    output: "{prefix}.sixframe.faa"
    benchmark: "benchmarks/{prefix}.sixframe.time"
    version: get_version("transeq")
    shell:
        "transeq -sequence {input} -clean -outseq {output} -table 11 -frame all"

rule compile_counts:
    """
    Use clade assignments and gene family assignments to compile a table of gene family counts by clade.

    input has switch to skip six frame translation for lastx or lastn

    dbfmt should be tbl.dbatch for fragmented dbs
    """
    input:
        "{prefix}.nocolon.annot.{taxdb}.lastx.{rank}.tsv",
        lambda w: \
            "{prefix}.nocolon{translate}.annot.gene_family.{db}.{dbfmt}.tsv"\
                .format(translate='.sixframe' \
                                if re.search(r'last[nx]',w.dbfmt) is None \
                                else '',
                        **w)
    output:
        "{prefix}.annot.{taxdb}.lastx.{rank}.vs.{db}.{dbfmt}.tsv"
    benchmark:
        "benchmarks/{prefix}.compile_counts.{taxdb}.lastx.{rank}.vs.{db}.{dbfmt}.time"
    log:
        "logs/{prefix}.compile_counts.{taxdb}.lastx.{rank}.vs.{db}.{dbfmt}.log"
    version:
        get_version('compile_hit_counts.py')
    params:
        three_column_opt="-L" if config.get('output_style','default').lower() == 'long' else ""
    shell:
        "compile_hit_counts.py {params.three_column_opt} -1 {input[0]} -2 {input[1]} -o {output} -S"

