"""
This workflow pulls out SSU rRNA genes using sortmeRNA, quality trims and joins pairs, then runs riboframe.

Structure:
+-.gitignore
+--README.md
+--LICENSE.md
+--config.yaml ssu_rna.snake
+--environment: workflows/test/conda/cDNA.yml
+--workflows: workflows/annotation/ssu_rna.snake 
+--scripts
|   +- python scripts: /tools/pymg/
|   +-
+- sortmeRNA hmms: /slipstream/home/jbryant/databases/
+--Snakefile

environment: source activate cDNA
example config file: config.cazy.yml (located /workflows/test/data/configs/config.cazy.yml

usage: snakemake -s ~/workflows/annotation/ssu_rna.snake --configfile ~/workflows/test/data/configs/ssu_rna.yml
"""

import os
import sys

snakefile_path = os.path.dirname(os.path.abspath(workflow.snakefile))[0:-11]
pymg_dir = os.path.sep.join([snakefile_path, 'tools', 'pymg'])

sys.path.append(snakefile_path)
sys.path.append(pymg_dir)
sortmerna_database_dir = "/slipstream/home/jbryant/databases/sortmerna/"


rule all:
    input: "ssu_rRNA/HSD00-20a-283-S2C012-0500.renamed.interleaved.noadapt.nophix.corrected.trimmed.dropse.rna.fastq"

rule sortmerna:
    """
    run sortmerna interleaved sequences
    """
    
    input:
        "/mnt/arginine/assemblies/Aloha_2_0/HSD00-20a/HSD00-20a-283-S2C012-0500-170428/HSD00-20a-283-S2C012-0500.renamed.interleaved.noadapt.nophix.corrected.trimmed.dropse.fastq"
    output:
        "ssu_rRNA/HSD00-20a-283-S2C012-0500.renamed.interleaved.noadapt.nophix.corrected.trimmed.dropse.rna.fastq"
    log:
        "logs/sortmerna.log"
    benchmark:
        "benchmarks/sortmerna.txt"
    shell:
        """
        echo 'made it here'
        sortmerna --ref {sortmerna_database_dir}silva-euk-18s-id95.fasta,{sortmerna_database_dir}silva-euk-18s-id95.db:{sortmerna_database_dir}silva-bac-16s-id90.fasta,{sortmerna_database_dir}silva-bac-16s-id90.db:{sortmerna_database_dir}silva-arc-16s-id95.fasta,{sortmerna_database_dir}silva-arc-16s-id95.db --reads {input} --aligned {output} --fastx --paired_in
        """
